SPOI-5053 APEX-56 #resolve #comment

ishark opened the pull request:
Fixing removal of terminated operators from physical plan when downanStream operators are also completed till shutdown window Id
Also fixed containers to be removed only when operators are removed from physical plan

tweise commented:
Some general comments:

"Persist" does not match implementation, the feature looks more like "mirror stream" and could be used for other purposes also.

Our yet to be published coding conventions mandate curly brace on new line for classes and methods. Also we don't want to wrap lines. Please adjust your IDE settings.

Does this implementation handle dynamic change of logical and physical plan?

ishark commented:
Have updated the code changes to adhere to formatting guidelines.
Yes, implementation handles dynamic re-partitioning.  Have added a test case for it.

Have not changed the name from persitStream to mirror yet.

tweise commented:
test this please

ishark commented:
Following test scenarios are covered with unit testing:

Basic tests:
1. Check persist operator is added to dag when persist is invoked on stream and dag validate succeeds
2. Check persist operator is added to dag when persist is invoked for specific sink
3. When stream for which persist is enabled is removed => the persist operator should also be removed
4. When an operator is removed, if persist is enabled for the operator => the persist operator should also be removed
5. When all sink operators are removed => the persist operator should also be removed

Functional validation:
1. persist is set for a specific sink without filtering => persist operator should receive identical input
2. persist is set for a specific sink with filtering => persist operator should receive filtered data
3. persist is set for stream without any filtering on sinks => identical data should be recieved
4. When persist is set for stream with filtering enabled for sinks => union of what flows to all the sinks should be persisted 
5. Case when persist operator itself is partitioned 
6. dynamic re partitioning of sink in stream being persisted => Tests that wrapper codec adjusts to new partitions
7. filtering with container local stream => tests that filtering works as expected

Negative tests:
1. Exception thrown when invalid input port type is used
2. Exception is thrown for invalid persist operator type

243826 commented on api/src/main/java/com/datatorrent/api/DAG.java:
```diff
@@ -106,6 +106,12 @@
 
     public StreamMeta addSink(Operator.InputPort<?> port);
 
+    public StreamMeta persistUsing(String name, Operator persistOperator, Operator.InputPort<?> persistOperatorInputPort);
```
can you document (java) the newly added methods?

243826 commented on engine/src/main/java/com/datatorrent/stram/plan/logical/DefaultKryoStreamCodec.java:
```diff
@@ -0,0 +1,60 @@
+package com.datatorrent.stram.plan.logical;
```
add apache 2.0 header here.

243826 commented on engine/src/main/java/com/datatorrent/stram/plan/logical/DefaultKryoStreamCodec.java:
```diff
@@ -0,0 +1,60 @@
+  @Override
+  public Object fromByteArray(Slice fragment)
+  {
+    ByteArrayInputStream is = new ByteArrayInputStream(fragment.buffer, fragment.offset, fragment.length);
```
although redundant here - any stream that's opened must be closed.

243826 commented on engine/src/main/java/com/datatorrent/stram/plan/logical/DefaultKryoStreamCodec.java:
```diff
@@ -0,0 +1,60 @@
+
+  private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException
+  {
+    in.defaultReadObject();
```
why is the Kryo being reinitialized here?

243826 commented on engine/src/main/java/com/datatorrent/stram/plan/logical/DefaultKryoStreamCodec.java:
```diff
@@ -0,0 +1,60 @@
+  @Override
+  public Slice toByteArray(T info)
+  {
+    ByteArrayOutputStream os = new ByteArrayOutputStream();
```
close missing?

243826 commented on engine/src/main/java/com/datatorrent/stram/stream/BufferServerSubscriber.java:
```diff
@@ -272,6 +290,7 @@ public Tuple sweep()
           Slice fm = polledFragments.pollUnsafe();
           com.datatorrent.bufferserver.packet.Tuple data = com.datatorrent.bufferserver.packet.Tuple.getTuple(fm.buffer, fm.offset, fm.length);
           Object o;
+          skipObject=false;
```
what happens if skipObject is reset to false only when it's true?

243826 commented on engine/src/main/java/com/datatorrent/stram/stream/PartitionAwareSink.java:
```diff
@@ -60,12 +62,20 @@ public void put(T payload)
 
+  protected boolean canPutPayloadToOutput(T payload)
+  {
+    if(partitions == null) {
```
unnecessary check?

243826 commented on engine/src/main/java/com/datatorrent/stram/stream/PartitionAwareSink.java:
```diff
@@ -15,9 +15,11 @@
  */
 package com.datatorrent.stram.stream;
 
+import com.datatorrent.stram.plan.logical.StreamCodecWrapperForPersistance;
```
unused import?

243826 commented on engine/src/main/java/com/datatorrent/stram/stream/BufferServerSubscriber.java:
```diff
@@ -350,5 +377,27 @@ public int getCount(boolean reset)
+    BufferReservoirForPersistStream(int capacity, StreamCodec<?> streamCodec)
+    {
+      super(capacity);
+      wrapperStreamCodec = (StreamCodecWrapperForPersistance<Object>) streamCodec;
```
can't you just modify the constructor so tht you receive StreamCodecWrapperForPersistance instead of generic StreamCodec<?>?

243826 commented on engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java:
```diff
@@ -371,6 +379,106 @@ public PhysicalPlan(LogicalPlan dag, PlanContext ctx) {
+              }
+            }
+
+            if (inputStreamCodecs.size() == 0) {
```
Remove this condition/block and replace with `if (!inputStreamCodecs.isEmpty())`.

243826 commented on engine/src/main/java/com/datatorrent/stram/stream/PartitionAwareSinkForPersistence.java:
```diff
@@ -0,0 +1,28 @@
+{
+  StreamCodecWrapperForPersistance<Object> serdeForPersistence;
+
+  public PartitionAwareSinkForPersistence(StreamCodec<T> serde, Set<Integer> partitions, int mask, Sink<T> output)
```
pass the correct type to the constructor in the first place?

243826 commented on engine/src/main/java/com/datatorrent/stram/stream/PartitionAwareSinkForPersistence.java:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  @Override
+  protected boolean canPutPayloadToOutput(T payload)
```
`canSendToOutput` as a name for this method?

243826 commented on engine/src/test/java/com/datatorrent/stram/plan/StreamPersistanceTests.java:
```diff
@@ -0,0 +1,942 @@
+import com.datatorrent.api.DAG.StreamMeta;
+import com.datatorrent.api.Partitioner.PartitionKeys;
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DefaultInputPort;
```
consolidate the imports using `import something.*;`

243826 commented on engine/src/test/java/com/datatorrent/stram/plan/StreamPersistanceTests.java:
```diff
@@ -0,0 +1,942 @@
+
+public class StreamPersistanceTests
+{
+  private static final Logger LOG = LoggerFactory.getLogger(StreamPersistanceTests.class);
```
logger. http://www.slf4j.org/faq.html#declaration_pattern

Commit:
SPOI-5053 APEX-56 #resolve #comment
Fixing removal of terminated operators from physical plan when downanStream operators are also completed till shutdown window Id
Also fixed containers to be removed only when operators are removed from physical plan

tweise commented on engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java:
```diff
@@ -1040,15 +1040,31 @@ public int processEvents()
               plan.removeTerminatedPartition(oper);
             }
             it.remove();
+          } else {
```
Entry is never removed from map. Please do that and plan.remove in a single place.

tweise commented on engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java:
```diff
@@ -762,18 +762,14 @@ public void testOperatorShutdown()
     mc1.sendHeartbeat();
     scm.monitorHeartbeat();
 
+    scm.monitorHeartbeat();
```
Why is this added and with no preceding assertion?

tweise commented on engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java:
```diff
@@ -1743,11 +1757,10 @@ public void run()
     ContainerHeartbeatResponse rsp = getHeartbeatResponse(sca);
 
-    if (containerIdle && isApplicationIdle()) {
+    if (heartbeat.getContainerStats().operators.isEmpty() && isApplicationIdle() && containerIdle) {
```
remove containerIdle, it is no longer needed
